src/query/ts/m3db/encoded_unconsolidated_step_iterator.go

```diff
@@ -0,0 +1,143 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	xts "github.com/m3db/m3/src/query/ts"
+)
+
+type encodedStepIterUnconsolidated struct {
+	mu          sync.RWMutex
+	lastBlock   bool
+	exhausted   bool
+	started     bool
+	meta        block.Metadata
+	validIters  []bool
+	seriesMeta  []block.SeriesMeta
+	seriesIters []encoding.SeriesIterator
+}
+
+func (b *encodedBlock) stepIterUnconsolidated() block.UnconsolidatedStepIter {
+	return &encodedStepIterUnconsolidated{
+		meta:        b.meta,
+		seriesMeta:  b.seriesMetas,
+		seriesIters: b.seriesBlockIterators,
+		lastBlock:   b.lastBlock,
+	}
+}
+
+type encodedUnconsolidatedStep struct {
+	time   time.Time
+	values []xts.Datapoints
+}
+
+func (s *encodedUnconsolidatedStep) Time() time.Time          { return s.time }
+func (s *encodedUnconsolidatedStep) Values() []xts.Datapoints { return s.values }
+
+func (it *encodedStepIterUnconsolidated) CurrentUnconsolidated() (
+	block.UnconsolidatedStep,
+	error,
+) {
+	it.mu.RLock()
+	if !it.started || it.exhausted {
+		it.mu.RUnlock()
+		panic("out of bounds")
+	}
+
+	var t time.Time
+	values := make([]xts.Datapoints, len(it.seriesIters))
+	for i, iter := range it.seriesIters {
+		// Skip this iterator if it's not valid
+		if !it.validIters[i] {
+			continue
+		}
+
+		dp, _, _ := iter.Current()
+		if i == 0 {
+			t = dp.Timestamp
+		}
+
+		values[i] = []xts.Datapoint{{
+			Timestamp: dp.Timestamp,
+			Value:     dp.Value,
+		}}
+	}
+
+	step := &encodedUnconsolidatedStep{
+		time:   t,
+		values: values,
+	}
+
+	it.mu.RUnlock()
+	return step, nil
+}
+
+func (it *encodedStepIterUnconsolidated) Next() bool {
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	if it.exhausted {
+		return false
+	}
+
+	if !it.started {
+		it.validIters = make([]bool, len(it.seriesIters))
+		it.started = true
+	}
+
+	var anyNext bool
+	for i, iter := range it.seriesIters {
+		if iter.Next() {
+			anyNext = true
+			it.validIters[i] = true
+		} else {
+			it.validIters[i] = false
+		}
+	}
+
+	if !anyNext {
+		it.exhausted = true
+	}
+
+	return anyNext
+
+}
+
+func (it *encodedStepIterUnconsolidated) StepCount() int {
+	panic("cannot request step count from unconsolidated step iterator")
```

Comment:
Seems dangerous, no? What guarantees do we have that this won't get called on this type?

Reply:
My thinking was that this is an invalid method to call on an unconsolidated step iterator, as we have no way of knowing step count; however this actually gives the projected step count post consolidation, so it's doable

Reply:
So it definitely makes sense to me that we can't effectively answer this question for this class; we have to iterate through the block in order to know how many steps it contains. Given that, my real question is whether we actually need this method. I worry about a case where we pass through a `encodedStepIterUnconsolidated` as the interface, call the method and then get a panic. 

I think there are two potential safe approaches here, depending on what's needed:

1. Get rid of this method from the interface (if that's possible).
2. If it's not, use upcasting to check when it's safe to use. That is,

```
type StepCounted interface {
    StepCount() int
}

// when using it
func Process(b block.Block) {

var stepCount int = someDefault
if sc, ok := b.(StepCounted); ok {
   stepCount = sc.StepCount()
}
}
```
Does that make sense?

Reply:
Sorry, my response may have been a bit inaccurate.

While we can't get the actual step count for unconsolidated series, `UnconsolidatedStepIterator.StepCount()` should return the expected step count for that step iterator, post-consolidation. As such it returns `bounds.StepSize()` to mirror how the other unconsolidated step iterator performs.

Reply:
If anything, might be a bit of a naming issue on that func

Reply:
Ok cool, so we can get rid of the panic then? That works for me.